dev/test/test_dataloader.c

```diff
@@ -0,0 +1,299 @@
+                    dataloader_next_batch(loader);
+                    // count up the tokens we see
+                    for (int i = 0; i < BT; i++) {
+                        int input_token = loader->inputs[i];
+                        int target_token = loader->targets[i];
+                        assert(input_token >= 0 && input_token < total_tokens);
+                        assert(target_token >= 0 && target_token < total_tokens);
+                        num_seen_inputs[input_token]++;
+                        num_seen_targets[target_token]++;
+                    }
+                    start += BT;
+                }
+            }
+        }
+    }
+
+    // verify that we saw all the tokens the correct number of times
+    int tokens_fit = batches_fit * (B * T * num_processes); // number of tokens that fit per shard
+    for (int s = 0; s < num_shards; s++) {
+        int start = s * num_tokens; // token id that starts this shard
+        // verify the inputs counts for this shard:
+        // - the first tokens_fit should have been seen num_epochs times
+        // - the rest of the tokens in that should should have been seen zero times
+        checkEquals(num_seen_inputs + start, tokens_fit, num_epochs);
+        checkEquals(num_seen_inputs + start + tokens_fit, num_tokens - tokens_fit, 0);
+        // verify the target counts. same thing but offset by 1
+        checkEquals(num_seen_targets + start + 1, tokens_fit, num_epochs);
+        checkEquals(num_seen_targets + start + 1 + tokens_fit, num_tokens - tokens_fit, 0);
+    }
+
+    // cleanup
+    for (int i = 0; i < num_processes; i++) {
+        dataloader_free(&loaders[i]);
+    }
+    free(num_seen_inputs);
+    free(num_seen_targets);
+    printf("OK\n");
+}
+
+int main(void) {
+
+    // generate a few dummy shards of data with incrementing tokens
+    int header[HEADER_SIZE];
+    uint16_t tokens[num_tokens];
+    for (int shard_id = 0; shard_id < num_shards; shard_id++) {
+        // ensure unique tokens across the shards for ez accounting below
+        int token_offset = shard_id * num_tokens;
+        for (int i = 0; i < num_tokens; i++) {
+            tokens[i] = token_offset + i;
+        }
+        // write the shard
+        snprintf(shard_name, 64, "shard_%04d.bin", shard_id);
+        header[0] = 20240520; // magic
+        header[1] = 1; // version
+        header[2] = num_tokens; // number of tokens within
+        FILE* shard_file = fopenCheck(shard_name, "wb");
+        fwrite(header, sizeof(int), HEADER_SIZE, shard_file);
+        fwrite(tokens, sizeof(uint16_t), num_tokens, shard_file);
+        fcloseCheck(shard_file);
+        printf("Wrote shard %s\n", shard_name);
+    }
+
+    test_simple();
+    test_multiprocess_simple();
+    test_shuffled();
+    test_multiprocess_shuffled();
+
+    // clean up the shards
+    for (int shard_id = 0; shard_id < num_shards; shard_id++) {
+        snprintf(shard_name, 64, "shard_%04d.bin", shard_id);
```

Comment:
nit: since we're reusing `64` for shard names here and above (`char shard_name[64];`) consider extracting into a global var and reusing - makes it easier to refactor later on.

Reply:
shard_name is a global buffer. instead of 64 -> sizeof (shard_name)